File: packages/shared-types/src/portfolio.ts
import type { TimeFrame } from './market';

export enum AssetType {
  STOCK = 'STOCK',
  ETF = 'ETF',
  MUTUAL_FUND = 'MUTUAL_FUND',
  BOND = 'BOND',
  CRYPTO = 'CRYPTO',
}

export interface Portfolio {
  id: string;
  userId: string;
  name: string;
  type: 'retirement' | 'personal' | 'taxable' | 'other';
  totalValue: number;
  totalCost: number;
  totalGain: number;
  totalGainPercent: number;
  dayChange: number;
  dayChangePercent: number;
  holdings: Holding[];
  currency: string;
  createdAt: Date;
  updatedAt: Date;
}

export interface Holding {
  id: string;
  portfolioId: string;
  symbol: string;
  name: string;
  assetType: AssetType;
  quantity: number;
  averageCost: number;
  currentPrice: number;
  totalValue: number;
  totalCost: number;
  gain: number;
  gainPercent: number;
  dayChange: number;
  dayChangePercent: number;
  taxLots: TaxLot[];
  addedAt: Date;
}

export interface TaxLot {
  id: string;
  holdingId: string;
  quantity: number;
  purchasePrice: number;
  purchaseDate: Date;
  gainLoss: number;
  gainLossPercent: number;
  holdingPeriod: 'short' | 'long';
}

export interface Watchlist {
  id: string;
  userId: string;
  name: string;
  items: WatchlistItem[];
}

export interface WatchlistItem {
  id: string;
  watchlistId: string;
  symbol: string;
  name: string;
  assetType: AssetType;
  currentPrice: number;
  dayChange: number;
  dayChangePercent: number;
  targetPrice?: number;
  alertEnabled: boolean;
  addedAt: Date;
}

export interface PortfolioAllocation {
  sector: string;
  value: number;
  percentage: number;
}

export interface PortfolioPerformance {
  portfolioId: string;
  period: TimeFrame;
  startValue: number;
  endValue: number;
  return: number;
  returnPercent: number;
  benchmark?: BenchmarkComparison;
}

export interface BenchmarkComparison {
  name: string;
  return: number;
  returnPercent: number;
}

File: apps/macos/OvaFlusDesktop/Models/Portfolio.swift
import Foundation

struct Holding: Identifiable, Codable, Hashable {
    let id: String
    let symbol: String
    let name: String
    let quantity: Double
    let avgCost: Double
    var currentPrice: Double

    var currentValue: Double { quantity * currentPrice }
    var totalCost: Double { quantity * avgCost }
    var gain: Double { currentValue - totalCost }
    var gainPercent: Double { totalCost > 0 ? (gain / totalCost) * 100 : 0 }
}

struct Portfolio: Identifiable, Codable {
    let id: String
    let name: String
    var holdings: [Holding]
    let createdAt: Date
    let updatedAt: Date

    var totalValue: Double { holdings.reduce(0) { $0 + $1.currentValue } }
    var totalGain: Double { holdings.reduce(0) { $0 + $1.gain } }
}
